java/card-bot/src/main/java/com/google/hangouts/chat/card/Bot.java

```diff
@@ -0,0 +1,209 @@
+    @PostMapping("/")
+    @ResponseBody
+    public Message onEvent(@RequestBody JsonNode event) {
+        Message reply = new Message();
+        String eventType = event.get("type").asText();
+        switch (eventType) {
+            case "ADDED_TO_SPACE":
+                JsonNode space = event.get("space");
+                String spaceType = space.get("type").asText();
+                if ("ROOM".equals(spaceType)) {
+                    String displayName = space.get("displayName").asText();
+                    String replyText = String.format("Thanks for adding me to %s", displayName);
+                    reply.setText(replyText);
+                } else {
+                    String displayName = event.get("user").get("displayName").asText();
+                    String replyText = String.format("Thanks for adding me to a DM, %s!", displayName);
+                    reply.setText(replyText);
+                }
+                break;
+            case "MESSAGE":
+                Card card = createCardResponse(event.get("message").get("text").asText());
+                reply.setCards(Collections.singletonList(card));
+                break;
+            case "CARD_CLICKED":
+                // Get the custom action name and custom parameter value out of the event object.
+                String actionName = event.get("action").get("actionMethodName").asText();
+                String customParameterValue = event.get("action").get("parameters").get(0).get("value").asText();
+                card = respondToInteractiveCardClick(actionName, customParameterValue);
+                reply.setCards(Collections.singletonList(card));
+                break;
+            case "REMOVED_FROM_SPACE":
+                String name = event.get("space").get("name").asText();
+                logger.info(String.format("Bot removed from %s", name));
+                break;
+            default:
+                reply.setText("Cannot determine event type");
+                break;
+        }
+        return reply;
+    }
+
+
+    /**
+     * Creates a card-formatted response based on the message sent in Hangouts Chat.
+     *
+     * See the reference for JSON keys and format for cards:
+     * https://developers.google.com/hangouts/chat/reference/message-formats/cards
+     *
+     * @param message the event object sent from Hangouts Chat
+     * @return a card responseStr as a JSON object node
+     */
+    private Card createCardResponse(String message) {
+        Card card = new Card();
+        List<WidgetMarkup> widgets = new ArrayList<>();
+        List<ActionParameter> customParameters = Collections.singletonList(
+                new ActionParameter().setKey(INTERACTIVE_BUTTON_KEY).setValue(message)
+        );
+
+        // Parse the message and add widgets to the responseStr in the order that
+        // they were requested in the message.
+        Stream.of(message.split(" ")).forEach((s -> {
+            if (s.contains("header")) {
```

Comment:
ooof this is hard to read, could you add helpers? like

if (s.contains("header")) {
     card.setHeader(createCardHeader());
}

Reply:
Mind if we defer this?

Yes, it is hard to read, but breaking things out in to lots of helper methods is also hard to read (harder to follow with additional layers of indirection.)

Might be better to change the overall approach and ditch dynamically building the card based on the user text -- just a simple bot that returns a card with the different types of widgets and that's all. But if we go this route, want to do that across the 3 implementations (python/node/java.)

Reply:
sure thing